### PR TITLE
added methods in mediawiki and download class for url

### LINKED ISSUE #1327: Methods in Mediawiki class to give an URL to retrieve given an articleId
We should have one per possible API (which allow to retrieve article content), like:
* `veApiArticleUrl(string: articleId)`
* `restApiArticleUrl(string: articleId)`
* ...

And use them wherever this makes sense.

The goal is to simplify the code and avoid many concatenation all over the code implying a kind of URL prefix and an articleId.

### PR DESCRIPTION
**Pull request by MananJethwani**
@kelson42 this resolves #1327

### CODE REVIEW

**Comment by kelson42** [comment]
@MananJethwani I have rebased on `master`

File: src/Downloader.ts [hunk 1]
```diff
@@ -232,6 +232,14 @@ class Downloader {
       await this.initLocalServices();
   }
 
+  public getBaseUrl(articleId: string): string{
+    return `${this.baseUrl}${encodeURIComponent(articleId)}`
+  }
+
+  public getBaseUrlForMainPage(articleId: string): string{
+    return `${this.baseUrlForMainPage}${encodeURIComponent(articleId)}`
+  }
```
Comment (kelson42) [inline src/Downloader.ts:241]: I understand the logic behind this change and this is not wrong. The problem I see is that the concept of "base url" is wrong. Let me try to explain the reason.

By letting the Downloader deciding how look the URLs we create a situation where it has to decide about the proper approach beetween what is available (Downloader), how the URL looks like (Mediawiki) and how they are treated (the Renderers). It fails to do that properly because most of the constraints are outside its control. This create the kind of incoherencies where a piece of code get an URL... but ultimatively it does not know what kind of URL it is (which is then hard/impossible to deal properly with).

Long story short, I believe we should keep this like it is... even wrong... we will have to change it soon anyway and this will imply its removal.

**Review by kelson42 — CHANGES_REQUESTED** [review]

**Comment by MananJethwani** [comment]
@kelson42  is it fine now?

File: src/Dump.ts [hunk 2]
```diff
@@ -203,7 +203,7 @@ export class Dump {
 
         /* Push Mediawiki:Offline.css (at the end) */
         // TODO: Weak URL (might fail in a number of cases where the wiki path is not like on Wikipedia)
-        const offlineCssUrl = `${downloader.mw.webUrl.href}?title=Mediawiki:offline.css&action=raw`;
+        const offlineCssUrl = downloader.mw.getWebArticleUrlRaw('?title=Mediawiki:offline.css&action=raw');
```
Comment (kelson42) [inline src/Dump.ts:206]: This should only take articleId in argument.

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/Dump.ts (hunk 2 above)
Comment (MananJethwani) [reply src/Dump.ts:206]: @kelson42 but there is no other place this is being used better to rename it as getWebQueryUrl ?????

File: src/MediaWiki.ts [hunk 3]
```diff
@@ -109,6 +109,30 @@ class MediaWiki {
...
+
+  public getApiQueryUrl(query: string): string {
+    return `${this.apiUrl.href}${query}`
+  }
+
+  public getWebArticleUrlRaw(article: string): string {
+    return `${this.webUrl.href}${encodeURIComponent(article)}`
+  }
+
+  public getWebQueryUrl(article: string): string {
```
Comment (kelson42) [inline src/MediaWiki.ts:132]: Variable shoud be articleid, not article

File: src/Dump.ts [hunk 4]
```diff
@@ -203,7 +203,7 @@ export class Dump {
 
         /* Push Mediawiki:Offline.css (at the end) */
         // TODO: Weak URL (might fail in a number of cases where the wiki path is not like on Wikipedia)
-        const offlineCssUrl = `${downloader.mw.webUrl.href}?title=Mediawiki:offline.css&action=raw`;
+        const offlineCssUrl = downloader.mw.getWebQueryUrl('?title=Mediawiki:offline.css&action=raw');
```
Comment (kelson42) [inline src/Dump.ts:206]: You should use getwebarticleraw()

File: src/MediaWiki.ts (hunk 3 above)
Comment (kelson42) [reply src/MediaWiki.ts:132]: Code is wrong, ideally this method should be based on getwebqueryurl()

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/MediaWiki.ts (hunk 3 above)
Comment (MananJethwani) [reply src/MediaWiki.ts:132]: I will remove this for now

**Review by kelson42 — APPROVED** [review]

File: src/MediaWiki.ts [hunk 5]
```diff
@@ -109,6 +109,30 @@ class MediaWiki {
...
+    return `${this.webUrl.href}${encodeURIComponent(article)}`
```
[the remark this answers is not in the document]
Comment (MananJethwani) [reply src/MediaWiki.ts]: ok

**Comment by MananJethwani** [comment]
@kelson42 is it fine now?.....or articleId should be encoded here?
